text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+├── examples/
+│   ├── simple.rs
+│   └── multi-file-example/
+│       ├── main.rs
+│       └── ex_module.rs
+└── tests/
+    ├── some-integration-tests.rs
+    └── multi-file-test/
+        ├── main.rs
+        └── test_module.rs
+```
+
+* `Cargo.toml` and `Cargo.lock` are stored in the root of your package (*package
+  root*).
+* Source code goes in the `src` directory.
+* The default library file is `src/lib.rs`.
+* The default executable file is `src/main.rs`.
+    * Other executables can be placed in `src/bin/`.
+* Benchmarks go in the `benches` directory.
+* Examples go in the `examples` directory.
+* Integration tests go in the `tests` directory.
+
+If a binary, example, bench, or integration test consists of multiple source
+files, place a `main.rs` file along with the extra [*modules*][def-module]
+within a subdirectory of the `src/bin`, `examples`, `benches`, or `tests`
+directory. The name of the executable will be the directory name.
+
+You can learn more about Rust's module system in [the book][book-modules].
+
+See [Configuring a target] for more details on manually configuring targets.
+See [Target auto-discovery] for more information on controlling how Cargo
+automatically infers target names.
+
+[book-modules]: https://doc.rust-lang.org/cargo/guide/../../book/ch07-00-managing-growing-projects-with-packages-crates-and-modules.html
+[Configuring a target]: https://doc.rust-lang.org/cargo/guide/../reference/cargo-targets.html#configuring-a-target
+[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package          '"package" (glossary entry)'
+[Target auto-discovery]: https://doc.rust-lang.org/cargo/guide/../reference/cargo-targets.html#target-auto-discovery
+[TOML]: https://toml.io/
+[crates.io]: https://crates.io/
+[SemVer]: https://semver.org
+[def-crate]:             https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#crate             '"crate" (glossary entry)'
+[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package           '"package" (glossary entry)'
+[def-package-registry]:  https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package-registry  '"package-registry" (glossary entry)'
+[def-manifest]:          https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#manifest          '"manifest" (glossary entry)'
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Single-file packages
+
+In addition to today's multi-file packages (`Cargo.toml` file with other `.rs`
+files),
+we are adding the concept of single-file packages which may contain an
+embedded manifest.
+There is no required way to distinguish a single-file
+`.rs` package from any other `.rs` file.
+
+A single-file package may contain an embedded manifest.
+An embedded manifest is stored using `TOML` in rust "frontmatter",
+a markdown code-fence with `cargo` at the start of the infostring at the top of
```

Comment:
I don't like that this makes a _language feature_ specific to Cargo. It should be possible to create alternative build tools to Cargo (similar to how JS can be used with `npm`, `yarn` or `pnpm`). The info string should not be part of the _language_ specification, nor the format (TOML).

Also, the correct term from CommonMark is a "fenced code block". A code fence is only the sequence of at least three `` ` `` or `~` characters that surround a fenced code block.

This document doesn't specify if the code fence syntax should conform to the CommonMark syntax, nor does it say what a code fence is. So it is unclear if rustc will accept this code:

`````md
   ~~~~cargo
   [package]
  ~~~~~~
`````

The [CommonMark spec](https://spec.commonmark.org/0.30/#fenced-code-blocks) would parse this as a valid fenced code block, because it says that the closing code fence must have _at least_ as many characters as the opening code fence, and the indentation of the opening and closing code fences need not match (although the indentation can't exceed 4 spaces).

Honestly, I don't think supporting all these edge cases in rustc is a good idea; that's also why I think using CommonMark syntax is a bad idea to begin with.

Reply:
This is a conversation for #3503